sparse/impl/KokkosSparse_spiluk_numeric_impl.hpp

```diff
@@ -21,644 +21,1000 @@
+    // }
+    return rv;
+  }
+
+  /**
+   * Common base class for SPILUK functors. Default version does not support
+   * blocks
+   */
+  template <class ARowMapType, class AEntriesType, class AValuesType,
+            class LRowMapType, class LEntriesType, class LValuesType,
+            class URowMapType, class UEntriesType, class UValuesType,
+            bool BlockEnabled>
+  struct Common {
+    ARowMapType A_row_map;
+    AEntriesType A_entries;
+    AValuesType A_values;
+    LRowMapType L_row_map;
+    LEntriesType L_entries;
+    LValuesType L_values;
+    URowMapType U_row_map;
+    UEntriesType U_entries;
+    UValuesType U_values;
+    LevelViewType level_idx;
+    WorkViewType iw;
+    lno_t lev_start;
+
+    Common(const ARowMapType &A_row_map_, const AEntriesType &A_entries_,
+           const AValuesType &A_values_, const LRowMapType &L_row_map_,
+           const LEntriesType &L_entries_, LValuesType &L_values_,
+           const URowMapType &U_row_map_, const UEntriesType &U_entries_,
+           UValuesType &U_values_, const LevelViewType &level_idx_,
+           WorkViewType &iw_, const lno_t &lev_start_,
+           const size_type &block_size_)
+        : A_row_map(A_row_map_),
+          A_entries(A_entries_),
+          A_values(A_values_),
+          L_row_map(L_row_map_),
+          L_entries(L_entries_),
+          L_values(L_values_),
+          U_row_map(U_row_map_),
+          U_entries(U_entries_),
+          U_values(U_values_),
+          level_idx(level_idx_),
+          iw(iw_),
+          lev_start(lev_start_) {
+      KK_REQUIRE_MSG(block_size_ == 0,
+                     "Tried to use blocks with the unblocked Common?");
+    }
+
+    KOKKOS_INLINE_FUNCTION
+    void init_scratch() const {}  // unblocked doesn't need scratch
+
+    // lset
+    KOKKOS_INLINE_FUNCTION
+    void lset(const size_type nnz, const scalar_t &value) const {
+      L_values(nnz) = value;
+    }
+
+    // uset
+    KOKKOS_INLINE_FUNCTION
+    void uset(const size_type nnz, const scalar_t &value) const {
+      U_values(nnz) = value;
+    }
+
+    // lset_id
+    KOKKOS_INLINE_FUNCTION
+    void lset_id(const size_type nnz) const { L_values(nnz) = scalar_t(1.0); }
+
+    KOKKOS_INLINE_FUNCTION
+    void lset_id(const member_type &team, const size_type nnz) const {
+      // Not sure a Kokkos::single is really needed here since the
+      // race is harmless
+      Kokkos::single(Kokkos::PerTeam(team),
+                     [&]() { L_values(nnz) = scalar_t(1.0); });
+    }
+
+    // divide. lhs /= rhs
+    KOKKOS_INLINE_FUNCTION
+    void divide(scalar_t &lhs, const scalar_t &rhs) const { lhs /= rhs; }
+
+    KOKKOS_INLINE_FUNCTION
+    void divide(const member_type &team, scalar_t &lhs,
+                const scalar_t &rhs) const {
+      Kokkos::single(
+          Kokkos::PerTeam(team),
+          [&](scalar_t &tmp) {
+            tmp = lhs / rhs;
+            lhs = tmp;
+          },
+          lhs);
```

Comment:
It is a bit odd when we have updated lhs (on device memory) in the lambda body then broadcast it among team too. I think we can do this:
 ```
 Kokkos::single(Kokkos::PerTeam(team), [&]() {lhs = lhs / rhs;});
 team.team_barrier();
```

Reply:
This code was in the original (before this PR), but I am happy to change it.